snapshot/btrfs/btrfs.go

```diff
@@ -7,51 +7,69 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/Sirupsen/logrus"
 	"github.com/containerd/btrfs"
 	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/mountinfo"
 	"github.com/containerd/containerd/plugin"
 	"github.com/containerd/containerd/snapshot"
 	"github.com/containerd/containerd/snapshot/storage"
 	"github.com/pkg/errors"
 )
 
-type btrfsConfig struct {
-	Device string `toml:"device"`
-}
-
 func init() {
 	plugin.Register("snapshot-btrfs", &plugin.Registration{
-		Type:   plugin.SnapshotPlugin,
-		Config: &btrfsConfig{},
+		Type: plugin.SnapshotPlugin,
 		Init: func(ic *plugin.InitContext) (interface{}, error) {
 			root := filepath.Join(ic.Root, "snapshot", "btrfs")
-			conf := ic.Config.(*btrfsConfig)
-			if conf.Device == "" {
-				// TODO: check device for root
-				return nil, errors.Errorf("btrfs requires \"device\" configuration")
-			}
-			return NewSnapshotter(conf.Device, root)
+			return NewSnapshotter(root)
 		},
 	})
 }
 
 type snapshotter struct {
-	device string // maybe we can resolve it with path?
+	device string // device of the root
 	root   string // root provides paths for internal storage.
 	ms     *storage.MetaStore
 }
 
+func getBtrfsDevice(root string, mounts []*mountinfo.Info) (string, error) {
+	device := ""
```

Comment:
device id is unreliable on btrfs
https://www.spinics.net/lists/linux-btrfs/msg58908.html